base/genimport/importer.go

```diff
@@ -291,99 +297,67 @@ func createPluginGoModFile(o *Output, pkgpath string, dir string) string {
-	goModDir, goModFile, err := findGoMod(localpath)
+func goModReplacementDirectives(o *Output, pkgModFileInfo modInfo) [][]byte {
+	m, err := getModuleFile(pkgModFileInfo)
 	if err != nil {
-		o.Errorf("error finding go.mod: %v", err)
+		o.Errorf("error getting go.mod", err)
 		return [][]byte{}
 	}
 
-	origGoModText, err := ioutil.ReadFile(goModFile)
-	if err != nil {
-		o.Errorf("error reading go.mod: %v", err)
-		return [][]byte{}
-	}
-	origLines := bytes.SplitAfter(origGoModText, []byte("\n"))
-
-	// Figure out current module name
-	// TODO: Could we use github.com/golang/mod to parse this?
-	var moduleName string
-	for _, line := range origLines {
-		if bytes.HasPrefix(line, []byte("module")) {
-			// TODO: is the module line guaranteed to have this syntax?
-			moduleName = string(bytes.Fields(line)[1])
-			break
-		}
-	}
-	if moduleName == "" {
-		o.Errorf("go.mod had no module line")
-	}
-
 	// Replace the current module with the right directory
 	replacementDirectives := [][]byte{
-		[]byte("replace "+moduleName+" => "+goModDir+"\n"),
+		[]byte("replace "+ pkgModFileInfo.Path+" => " + pkgModFileInfo.Dir + "\n"),
 	}
 
-	// Copy over the replaces, since they only apply when in the main module
-	for _, line := range origLines {
-		// TODO: this doesn't handle block syntax e.g. replace ( ... )
-		// TODO: this doesn't work if the replace directives found are themselves localized
-		if bytes.HasPrefix(line, []byte("replace")) {
-			replacementDirectives = append(replacementDirectives, line)
-		}
+	for _, repplaceDirective := range m.Replace {
+		// TODO: this doesn't work if the replace directives found are themselves local to the repo
+		replacementDirectives = append(replacementDirectives,
+			[]byte("replace " + repplaceDirective.Old.Path + " => " +
+				repplaceDirective.New.Path + " " + repplaceDirective.New.Version + "\n"))
 	}
+
 	return replacementDirectives
 }
 
-var noGoModOnPath = errors.New("no go.mod file on path")
+type modInfo struct {
+	Path      string `json:"Path"`
+	Dir       string `json:"Dir"`
+	GoMod     string `json:"GoMod"`
+	GoVersion string `json:"GoVersion"`
+	Main      bool   `json:"Main"`
+}
 
-func findGoMod(d string) (string, string, error) {
-	dir := d
-	for {
-		if rel, err := filepath.Rel("/", dir); err != nil || rel == "." {
-			break
-		}
-		maybeGoMod := filepath.Join(dir, "go.mod")
-		if exists(maybeGoMod) {
-			return dir, maybeGoMod, nil
-		}
-		dir = filepath.Dir(dir)
+func getModuleFile(i modInfo) (*modfile.File, error) {
+	raw, err := ioutil.ReadFile(i.GoMod)
+	if err != nil {
+		return nil, fmt.Errorf("reading go.mod file: %w", err)
 	}
-	return "", "", noGoModOnPath
+
+	return modfile.Parse("go.mod", raw, nil)
 }
 
-func exists(name string) bool {
-	if _, err := os.Stat(name); err != nil {
-		if os.IsNotExist(err) {
-			return false
-		}
+func getModuleFileInfo(dir string) (modInfo, error) {
+	cmd := exec.Command("go", "list", "-m", "-json", "-f", "{{.GoMod}}")
```

Comment:
```suggestion
	// https://github.com/golang/go/issues/44753#issuecomment-790089020
	cmd := exec.Command("go", "list", "-m", "-json", "-f", "{{.GoMod}}")
```
It might be nice to retain a comment tracking the progress on exposing this information directly via an API.